drivers/pinctrl/pinctrl_npcx.c

```diff
@@ -0,0 +1,226 @@
+		return true;
+	}
+#elif defined(CONFIG_SOC_SERIES_NPCX9)
+	if (group == 0x00 || (group >= 0x02 && group <= 0x06) || group == 0x0b ||
+		group == 0x0d || (group >= 0x0f && group <= 0x12)) {
+		return true;
+	}
+#endif
+	return false;
+}
+
+static void npcx_periph_pinmux_configure(const struct npcx_periph *alt, bool func,
+	bool is_locked)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+	uint8_t alt_mask = BIT(alt->bit);
+
+	/*
+	 * func == 0 means select GPIO, otherwise Alternate Func.
+	 * inverted == 0:
+	 *    Set devalt bit to select Alternate Func.
+	 * inverted == 1:
+	 *    Clear devalt bit to select Alternate Func.
+	 */
+	if (func != alt->inverted) {
+		NPCX_DEVALT(scfg_base, alt->group) |=  alt_mask;
+	} else {
+		NPCX_DEVALT(scfg_base, alt->group) &= ~alt_mask;
+	}
+
+	if (is_locked && npcx_periph_pinmux_has_lock(alt->group)) {
+		NPCX_DEVALT_LK(scfg_base, alt->group) |= alt_mask;
+	}
+}
+
+static void npcx_periph_pupd_configure(const struct npcx_periph *pupd,
+	enum npcx_io_bias_type type)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
+
+	if (type == NPCX_BIAS_TYPE_NONE) {
+		NPCX_PUPD_EN(scfg_base, pupd->group) &= ~BIT(pupd->bit);
+	} else {
+		NPCX_PUPD_EN(scfg_base, pupd->group) |=  BIT(pupd->bit);
+	}
+}
+
+static void npcx_periph_configure(const pinctrl_soc_pin_t *pin)
+{
+	if (pin->cfg.periph.type == NPCX_PINCTRL_TYPE_PERIPH_PINMUX) {
+		/* Configure peripheral device's pinmux functionality */
+		npcx_periph_pinmux_configure(&pin->cfg.periph,
+					     !pin->flags.pinmux_gpio,
+					     pin->flags.pinmux_lock);
+	} else if (pin->cfg.periph.type == NPCX_PINCTRL_TYPE_PERIPH_PUPD) {
+		/* Configure peripheral device's internal PU/PD */
+		npcx_periph_pupd_configure(&pin->cfg.periph,
+			pin->flags.io_bias_type);
+	}
+}
+
+/* Pin-control local functions for IO pads */
+static void npcx_iopad_configure(const struct npcx_io_pad *pad, gpio_flags_t flags)
+{
+	gpio_pin_configure(npcx_get_gpio_dev(pad->port), pad->bit, flags);
+}
+
+static int npcx_io_pad_configure(const pinctrl_soc_pin_t *pin)
+{
+	if (pin->flags.type == NPCX_PINCTRL_TYPE_IO_PAD) {
+		gpio_flags_t flags = 0;
+
+		/* Enable low-voltage input detection? */
+		if (pin->flags.io_1p8v_input) {
+			flags |= GPIO_VOLTAGE_1P8;
+		}
+
+		/* Pin types & init states */
+		if (pin->flags.io_type == NPCX_IO_TYPE_OUTPUT_LOW) {
+			flags |= GPIO_OUTPUT_LOW;
+		} else if (pin->flags.io_type == NPCX_IO_TYPE_OUTPUT_HIGH) {
+			flags |= GPIO_OUTPUT_HIGH;
+		} else {
+			flags |= GPIO_INPUT;
+		}
+
+		/* Pin drive types */
+		if (pin->flags.io_drive_type == NPCX_DRIVE_TYPE_OPEN_DRAIN) {
+			flags |= GPIO_OPEN_DRAIN;
+		} else {
+			flags |= GPIO_PUSH_PULL;
+		}
+
+		/* Pin bias types */
+		if (pin->flags.io_bias_type == NPCX_BIAS_TYPE_PULL_DOWN) {
+			flags |= GPIO_PULL_DOWN;
+		} else if (pin->flags.io_bias_type == NPCX_BIAS_TYPE_PULL_UP) {
+			flags |= GPIO_PULL_UP;
+		}
+
+		/* Configure IO pad */
+		npcx_iopad_configure(&pin->cfg.io_pad, flags);
```

Comment:
Call `gpio_pin_configure()` here directly.  The function names `npcx_iopad_configure()` and `npcx_io_pad_configure()` are so similar, I thought this was a recursive call.

Reply:
Done.